Fix a number of issues related to runtime resolution of methods

dmitry-ten opened the pull request:
The client-side handler of `ResolvedMethod_getMultipleResolvedMethods`
did not check the value of `J9JIT_RUNTIME_RESOLVE` flag for some of
the methods it created, which resulted in bugs when `-Xjit:rtResolve`
option is specified.

This commit refactors `ResolvedMethod_getMultipleResolvedMethods`
and `TR_ResolvedJ9JITServerMethod::cacheResolvedMethodCallees` to
call resolved method creation routine on the client directly, instead
of copy-pasting the code from that routine, which fixes the problem.

The code here is a small chunk of PR #12147, which should fully refactor
resolved methods creation, when it is merged at a later date.

dmitry-ten commented:
@mpirvu ready for review. This fixes a crash in `DaaLoadTest_daa2_special_5m_20` for JDK8 on Power, and potentially fixes some other segfaults as well, but I'm in the process of testing some of those.

mpirvu reviewed (approved):
LGTM

mpirvu commented:
jenkins test sanity,extended,sanity.system,extended.system plinuxjit,xlinuxjit,zlinuxjit jdk11

mpirvu commented:
sanity.functional failed `CondyPrimitie_1` test on all 3 Linux platforms.
Is this a new test? Does it pass without JITServer?

extended.system failed two SCC tests on pLinux:
` SharedClasses.SCM23.MultiCL_1`
```
variation: Mode610
JVM_OPTIONS: -XX:+UseJITServer -Xcompressedrefs -Xjit -Xgcpolicy:gencon 
MCL3 stderr Unhandled exception
MCL3 stderr Type=Segmentation error vmState=0x00020011
...
```

`SharedClasses.SCM23.MultiThreadMultiCL_1`
```
variation: Mode610
JVM_OPTIONS: -XX:+UseJITServer -Xcompressedrefs -Xjit -Xgcpolicy:gencon 
MTM1 stderr Type=Segmentation error vmState=0x00020011
```

dmitry-ten commented:
`SharedClasses.SCM23.MultiThreadMultiCL_1`: Segfaults  are the same issue as in https://github.com/eclipse-openj9/openj9/issues/11185,  this PR doesn't fix them, unfortunately. 

`CondyPrimitive_1`: it's not a new test, non-JITServer nightly runs don't fail it, and the test has `-Xjit:rtResolve`, so this failure should be investigated.

dmitry-ten commented:
@mpirvu, ready for review.
I split this PR into 3 commits as each addresses a different issue. With these changes, both `SharedClasses.SCM23.MultiThreadMultiCL_1` and `CondyPrimitive_1` will pass.

dmitry-ten commented:
Actually, I still see occasional crashes in `SharedClasses.SCM23.MultiThreadMultiCL_1`, but the crash is identical to #11185, i.e. corrupted `RuntimeAssumptionsTable`, which is unrelated to `rtResolve` issues, so this PR should still be good to go.

mpirvu commented:
jenkins test sanity,extended,sanity.system,extended.system plinuxjit,xlinuxjit,zlinuxjit jdk11

dmitry-ten commented on runtime/compiler/env/VMJ9Server.cpp:
```diff
@@ -166,7 +166,7 @@ TR_J9ServerVM::createResolvedMethodWithSignature(TR_Memory * trMemory, TR_Opaque
    {
    TR_ResolvedJ9Method *result = NULL;
-   if (_compInfoPT->getCompilation()->compileRelocatableCode())
+   if (isAOT_DEPRECATED_DO_NOT_USE())
```
`compileRelocatableCode` requires compilation object to exist, but this method will be called before it's initialized if we are creating a resolved method for the method being compiled.

mpirvu commented:
On zLinux there is a failure for extended.functional: cmdLineTester_pltest_j9sig_ext_0
On pLinux we have two failures from extended.system suite: SharedClasses.SCM23.MultiCL_1, SharedClasses.SCM23.MultiThreadMultiCL_1. These are known.

mpirvu commented:
In the nightlies I am not seeing `cmdLineTester_pltest_j9sig_ext_0` being done for zLinux for Java11, while for Java8 it passes.
@dmitry-ten could you please double check that this PR does not induce a failure for `cmdLineTester_pltest_j9sig_ext_0` on zLinux? Thanks

dmitry-ten commented:
Error message from the test:
```
[2021-06-09T06:07:30.326Z]  [ERR] 1: j9mmap_test9
[2021-06-09T06:07:30.326Z]  [ERR] 	/home/jenkins/workspace/Build_JDK11_s390x_linux_jit_Personal/openj9/runtime/tests/port/j9mmapTest.c line 1541: Modified string does not match in updated buffer
[2021-06-09T06:07:30.326Z]  [ERR] 
[2021-06-09T06:07:30.326Z]  [ERR] 		LastErrorNumber: -108
[2021-06-09T06:07:30.326Z]  [ERR] 		LastErrorMessage: No such file or directory
[2021-06-09T06:07:30.326Z]  [ERR] 
[2021-06-09T06:07:30.326Z]  [ERR] 2: j9mmap_test10
[2021-06-09T06:07:30.326Z]  [ERR] 	/home/jenkins/workspace/Build_JDK11_s390x_linux_jit_Personal/openj9/runtime/tests/port/j9mmapTest.c line 1760: Modified string does not match in updated buffer
[2021-06-09T06:07:30.326Z]  [ERR] 
[2021-06-09T06:07:30.326Z]  [ERR] 		LastErrorNumber: -108
[2021-06-09T06:07:30.326Z]  [ERR] 		LastErrorMessage: No such file or directory
[2021-06-09T06:07:30.326Z]  [ERR] 
```
It's due to being unable to open a file, which seems like it couldn't have been caused by this change. I'm re-running the tests to confirm

mpirvu commented:
jenkins test sanity,extended,sanity.system,extended.system plinuxjit,xlinuxjit,zlinuxjit jdk11

mpirvu commented:
Sanity.functional on pLinux/xlinux shows
```
===============================================
Running test jit_jitt_array_compress_3 ...
===============================================
jit_jitt_array_compress_3 Start Time: Thu Jun 10 18:55:43 2021 Epoch Time (ms): 1623365743739
variation: -Xjit:disableAsyncCompilation,count=1 -XXgc:forcedShiftingCompressionAmount=4
```
I wonder if this is a new thing, not related to JITServer

mpirvu commented:
jenkins test sanity plinux,xlinux jdk11

dmitry-ten commented:
Consdering that other recent JITServer PRs have the same failure, I think it's safe to say that it's not related to this PR.

dmitry-ten commented:
I checked with latest master, and the crash still happens, but only with JITServer, so we do have a new bug.

mpirvu commented:
> I checked with latest master, and the crash still happens, but only with JITServer, so we do have a new bug.

Thanks!
Two things to check: (1) was there something non-jitserver related that was delivered recently and broke JITServer? I am seeing this failure in the nightly tests for last nigh, but in the day before.
(2) Is this a new test?
